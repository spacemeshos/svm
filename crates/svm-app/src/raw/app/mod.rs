//!                   `Spawn-App` Raw Format
//!  ----------------------------------------------------------
//!  |               |                                        |
//!  |    `version`  |        `AppTemplate` (`Address`)       |
//!  |_______________|________________________________________|
//!  |            |                                           |
//!  | ctor index |            ctor-buf (`func-buf`)          |
//!  |____________|___________________________________________|
//!  |                                                        |
//!  |                 ctor-args (`func-args``)               |
//!  |________________________________________________________|
//!
//!

mod serialize;
mod wire;

pub use serialize::{DefaultAppDeserializer, DefaultAppSerializer};
pub use wire::{decode_spawn_app, encode_spawn_app};
