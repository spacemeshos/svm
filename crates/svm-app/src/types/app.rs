use std::fmt;

use serde::{Deserialize, Serialize};
use svm_common::Address;

#[derive(Serialize, Deserialize, PartialEq)]
pub struct App {
    pub template: Address,
    pub creator: Address,
}

impl fmt::Debug for App {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let template = self.fmt_template(&self.template);
        let creator = self.fmt_creator(&self.creator);

        let msg = [template, creator].join("\n");
        write!(f, "{}", msg)
    }
}

impl App {
    #[inline]
    fn fmt_template(&self, addr: &Address) -> String {
        format!("Template: {}", self.fmt_address(addr))
    }

    #[inline]
    fn fmt_creator(&self, addr: &Address) -> String {
        format!("Creator: {}", self.fmt_address(addr))
    }

    #[inline]
    fn fmt_address(&self, addr: &Address) -> String {
        addr.fmt(4, 4, " ")
    }
}
