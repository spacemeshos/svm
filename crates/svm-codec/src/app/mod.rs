//              Spawn-App Raw Format
//
//  ----------------------------------------------------------
//  |             |                                          |
//  |  `version`  |        `AppTemplate` (`Address`)         |
//  |_____________|__________________________________________|
//  |               |                                        |
//  | ctor (String) |           ctor `CallData`              |
//  |_______________|________________________________________|
//

mod raw;
mod serialize;

pub use raw::{decode_spawn_app, encode_spawn_app};
pub use serialize::{DefaultAppDeserializer, DefaultAppSerializer};
