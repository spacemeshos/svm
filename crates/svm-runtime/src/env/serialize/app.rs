use std::io::Cursor;

use svm_codec::{ReadExt, WriteExt};
use svm_types::{App, SpawnerAddr, TemplateAddr};

use crate::env::{traits, ExtApp};

use traits::{AppDeserializer, AppSerializer};

/// Default serializer for `App`
pub struct DefaultAppSerializer;

/// Default deserializer for `App`
pub struct DefaultAppDeserializer;

impl AppSerializer for DefaultAppSerializer {
    fn serialize(app: &ExtApp) -> Vec<u8> {
        let mut w = Vec::new();

        encode_template(app, &mut w);
        encode_name(app, &mut w);
        encode_spawner(app, &mut w);

        w
    }
}

fn encode_template(app: &ExtApp, w: &mut Vec<u8>) {
    let addr = app.template_addr();

    w.write_address(addr.inner());
}

fn encode_name(app: &ExtApp, w: &mut Vec<u8>) {
    w.write_string(app.name());
}

fn encode_spawner(app: &ExtApp, w: &mut Vec<u8>) {
    let spawner = app.spawner();

    w.write_address(spawner.inner());
}

impl AppDeserializer for DefaultAppDeserializer {
    fn deserialize(bytes: &[u8]) -> Option<ExtApp> {
        let mut cursor = Cursor::new(bytes);

        let template = match cursor.read_address() {
            Ok(addr) => TemplateAddr::new(addr),
            _ => return None,
        };

        let name = match cursor.read_string() {
            Ok(Ok(name)) => name,
            _ => return None,
        };

        let spawner = match cursor.read_address() {
            Ok(addr) => SpawnerAddr::new(addr),
            _ => return None,
        };

        let base = App::new(template, name);
        let app = ExtApp::new(&base, &spawner);

        Some(app)
    }

    fn desrialize_template_addr(bytes: &[u8]) -> Option<TemplateAddr> {
        let mut cursor = Cursor::new(bytes);

        cursor.read_address().ok().map(|addr| addr.into())
    }
}
