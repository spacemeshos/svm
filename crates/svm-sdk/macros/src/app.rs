use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use serde_json::Value;

use syn::{Error, Item, ItemMod, ItemStruct, ItemType, ItemUse, Result};

use super::{function, r#struct};
use crate::{api, schema, Function, Schema, Struct};

use r#function::{func_attrs, has_default_fundable_hook_attr};
use r#struct::has_storage_attr;

pub struct App {
    name: Ident,
    functions: Vec<Function>,
    structs: Vec<Struct>,
    imports: Vec<ItemUse>,
    aliases: Vec<ItemType>,
    default_fundable_hook: Option<Ident>,
}

impl App {
    pub fn name(&self) -> String {
        self.name.to_string()
    }

    pub fn functions(&self) -> &[Function] {
        &self.functions
    }

    pub fn structs(&self) -> &[Struct] {
        &self.structs
    }

    pub fn imports(&self) -> &[ItemUse] {
        &self.imports
    }

    pub fn aliases(&self) -> &[ItemType] {
        &self.aliases
    }

    pub fn default_fundable_hook(&self) -> Option<Ident> {
        self.default_fundable_hook.clone()
    }

    pub fn set_default_fundable_hook(&mut self, hook: Ident) {
        self.default_fundable_hook = Some(hook)
    }
}

pub fn expand(_args: TokenStream, input: TokenStream) -> Result<(Schema, TokenStream)> {
    let module = syn::parse2(input)?;
    let app = parse_app(module)?;
    let schema = schema::app_schema(&app)?;

    let imports = app.imports();
    let aliases = app.aliases();

    let structs = expand_structs(&app)?;
    let functions = expand_functions(&app)?;
    let alloc_func = alloc_func_ast();

    #[cfg(feature = "api")]
    let api = api::json_api(&schema);

    #[cfg(feature = "api")]
    let stream = api::json_tokenstream(&api);

    #[cfg(feature = "api")]
    let data = api::json_data_layout(&schema);

    write_schema(&app, &api, &data);

    let ast = quote! {
        // #(#imports)*

        // #(#aliases)*

        #alloc_func

        #structs

        #functions

        #[cfg(all(feature = "api", not(target_arch = "wasm32")))]
        pub fn raw_schema() -> String {
            #stream.to_string()
        }
    };

    Ok((schema, ast))
}

pub fn parse_app(mut raw_app: ItemMod) -> Result<App> {
    let name = raw_app.ident.clone();

    let mut functions = Vec::new();
    let mut structs = Vec::new();
    let mut imports = Vec::new();
    let mut aliases = Vec::new();

    let (_, content) = raw_app.content.take().unwrap();

    for item in content {
        // TODO: Is is possible to extract the `item` real `Span`?
        let span = Span::call_site();

        match item {
            Item::Fn(item) => {
                let func = Function::new(item, functions.len());
                functions.push(func);
            }
            Item::Struct(item) => {
                let strukt = Struct::new(item);
                structs.push(strukt);
            }
            Item::Use(item) => imports.push(item),
            Item::Type(item) => aliases.push(item),
            Item::Const(..) => {
                let msg = "declaring `const` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Enum(..) => {
                let msg = "declaring `enum` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::ExternCrate(..) => {
                let msg = "using `extern crate` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::ForeignMod(..) => {
                let msg =
                    "using foreign items such as `extern \"C\"` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Impl(..) => {
                let msg = "using `impl` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Macro(..) => {
                let msg = "declaring `macro_rules!` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Macro2(..) => {
                let msg = "declaring `macro` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Mod(..) => {
                let msg = "declaring new modules inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Static(..) => {
                let msg = "declaring new `static` items inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Trait(..) => {
                let msg = "declaring new traits inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::TraitAlias(..) => {
                let msg = "using trait aliases inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Union(..) => {
                let msg = "declaring `union` inside `#[app]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Verbatim(item) => {
                let msg = format!("invalid Rust code: {}", item);
                return Err(Error::new(span, msg));
            }
            Item::__Nonexhaustive => unreachable!(),
        }
    }

    let mut app = App {
        name,
        functions,
        structs,
        imports,
        aliases,
        default_fundable_hook: None,
    };

    let default = extract_default_fundable_hook(&app)?;

    if default.is_some() {
        app.set_default_fundable_hook(default.unwrap());
    }

    Ok(app)
}

fn extract_default_fundable_hook(app: &App) -> Result<Option<Ident>> {
    let span = Span::call_site();
    let mut seen_default_fundable_hook = false;
    let mut default = None;

    for func in app.functions().iter() {
        let attrs = func_attrs(func).unwrap();

        if has_default_fundable_hook_attr(&attrs) {
            if seen_default_fundable_hook {
                return Err(Error::new(
                    span,
                    "There can be only a single default `fundable hook`",
                ));
            }

            seen_default_fundable_hook = true;

            default = Some(func.raw_name());
        }
    }

    Ok(default)
}

#[cfg(all(feature = "api", target_arch = "wasm32"))]
fn write_schema(app: &App, api: &Value, data: &Value) {
    api::json_write(&format!("{}-api.json", app.name()), api);
    api::json_write(&format!("{}-data.json", app.name()), data);
}

#[cfg(any(not(feature = "api"), not(target_arch = "wasm32")))]
fn write_schema(app: &App, api: &Value, data: &Value) {
    //
}

fn expand_structs(app: &App) -> Result<TokenStream> {
    let mut structs = Vec::new();

    validate_structs(app)?;

    for strukt in app.structs() {
        let strukt = r#struct::expand(strukt)?;

        structs.push(strukt);
    }

    let ast = quote! {
        #(#structs)*
    };

    Ok(ast)
}

fn validate_structs(app: &App) -> Result<()> {
    let mut seen_storage = false;

    for strukt in app.structs() {
        match strukt.attrs() {
            Ok(attrs) => {
                if has_storage_attr(attrs) {
                    if seen_storage {
                        let msg = format!("an App can have only a single `#[storage]`");
                        let span = Span::call_site();

                        return Err(Error::new(span, msg));
                    }

                    seen_storage = true;
                }
            }
            Err(err) => return Err(err.clone()),
        }
    }

    Ok(())
}

fn expand_functions(app: &App) -> Result<TokenStream> {
    validate_funcs(app)?;

    let mut funcs = Vec::new();

    for func in app.functions() {
        let func = function::expand(func, app)?;

        funcs.push(func);
    }

    let implicit_fundable_hook = if app.default_fundable_hook().is_some() {
        quote! {}
    } else {
        function::fundable_hook::expand_default()?
    };

    let ast = quote! {
        #(#funcs)*

        #implicit_fundable_hook
    };

    Ok(ast)
}

fn validate_funcs(app: &App) -> Result<()> {
    Ok(())
}

fn alloc_func_ast() -> TokenStream {
    quote! {
        // injects the `#[global_allocator]`
        extern crate svm_sdk;

        #[no_mangle]
        pub extern "C" fn svm_alloc(size: u32) -> u32 {
            let ptr = svm_sdk::alloc(size as usize);

            ptr.offset() as u32
        }
    }
}
