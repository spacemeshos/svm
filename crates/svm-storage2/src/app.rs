use std::collections::HashMap;
use std::rc::Rc;

use crate::{layout::DataLayout, raw::RawStorage};

pub struct AppStorage {
    raw_storage: Rc<RawStorage>,

    layout: DataLayout,

    uncommitted: HashMap<u32, Vec<u8>>,
}

impl AppStorage {
    pub fn new(layout: DataLayout, raw_storage: Rc<RawStorage>) -> Self {
        Self {
            layout,
            raw_storage,
            uncommitted: HashMap::new(),
        }
    }

    pub fn read_var(&self, var_id: u32) -> Vec<u8> {
        let var = self.uncommitted.get(&var_id).map(|v| v.to_owned());

        var.unwrap_or_else(|| {
            let (off, len) = self.var_layout(var_id);

            self.raw_storage.read(off, len)
        })
    }

    pub fn write_var(&mut self, var_id: u32, value: Vec<u8>) {
        self.uncommitted.insert(var_id, value);
    }

    pub fn commit(&mut self) {
        // ....

        self.uncommitted.clear();
    }

    #[inline]
    fn var_layout(&self, var_id: u32) -> (u32, u32) {
        self.layout.get_var(var_id)
    }
}
