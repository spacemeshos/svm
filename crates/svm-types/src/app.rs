use std::fmt;

use crate::{Address, TemplateAddr};

/// An in-memory representation of an app.
#[derive(PartialEq)]
pub struct App {
    /// `App` version.
    pub version: u32,

    /// `Address` of the `AppTemplate`, the App is being spawned from.
    pub template: TemplateAddr,
}

impl fmt::Debug for App {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let version = self.fmt_version(self.version);
        let template = self.fmt_template(&self.template);

        let msg = [version, template].join("\n");

        writeln!(f, "{}", msg)
    }
}

impl App {
    fn fmt_version(&self, ver: u32) -> String {
        format!("Version: {}", ver)
    }

    fn fmt_template(&self, addr: &TemplateAddr) -> String {
        format!("Template: {}", self.fmt_address(addr.inner()))
    }

    fn fmt_address(&self, addr: &Address) -> String {
        addr.fmt(4, 4, " ")
    }
}
