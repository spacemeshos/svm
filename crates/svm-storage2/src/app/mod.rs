use std::collections::HashMap;

mod raw;
use raw::{RawChange, RawStorage};

mod kv;
pub use kv::AppKVStore;

use crate::layout::{DataLayout, VarId};

///
/// The `AppStorage` manages a running app's storage.
///
/// While an app is running it performs read and write operations.
///
/// Reads operations don't modify the app's storage.
/// Write operations save changes to be commited after an app's execution succeeded.
///
/// Reading a variable will return the most updated value of the variable.
///
/// If app's execution fails - no changes will be persisted (`commit` won't be called).
///
pub struct AppStorage {
    /// Interface to the raw storage (key-value wrapper)
    raw_storage: RawStorage,

    /// App Fixed-Sized variables layout
    layout: DataLayout,

    /// Uncommited changes
    uncommitted: HashMap<VarId, Vec<u8>>,
}

impl AppStorage {
    /// New instance for managing app's variabled specified by `layout`.
    /// App's storage is backed by key-value store `kv`.
    pub fn new(layout: DataLayout, app_kv: AppKVStore) -> Self {
        Self {
            layout,
            raw_storage: RawStorage::new(app_kv, 32),
            uncommitted: HashMap::new(),
        }
    }

    /// Reads variable `var_id`.
    pub fn read_var(&self, var_id: VarId) -> Vec<u8> {
        let var = self.uncommitted.get(&var_id).cloned();

        var.unwrap_or_else(|| {
            let (off, len) = self.var_layout(var_id);

            let bytes = self.raw_storage.read(off, len);

            debug_assert_eq!(bytes.len(), len as usize);

            bytes
        })
    }

    /// Marks variable as `dirty`. Upon `commit` will persist the variable.
    pub fn write_var(&mut self, var_id: VarId, value: Vec<u8>) {
        let (_off, len) = self.var_layout(var_id);

        debug_assert_eq!(value.len(), len as usize);

        self.uncommitted.insert(var_id, value);
    }

    /// Returns the layout of variable `var_id`.
    /// The layout is a tuple of `(offset, length)`.
    #[inline]
    pub fn var_layout(&self, var_id: VarId) -> (u32, u32) {
        self.layout.get_var(var_id)
    }

    /// Commits modified (a.k.a) variables into the raw storage.
    pub fn commit(&mut self) {
        let var_offset: HashMap<VarId, u32> = self
            .uncommitted
            .keys()
            .map(|var_id| {
                let (off, _len) = self.var_layout(*var_id);

                (*var_id, off)
            })
            .collect();

        let changes = self
            .uncommitted
            .drain()
            .map(|(var_id, data)| {
                let offset = *var_offset.get(&var_id).unwrap();

                RawChange { offset, data }
            })
            .collect::<Vec<_>>();

        self.raw_storage.write(&changes);

        debug_assert!(self.uncommitted.is_empty());
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::rc::Rc;
    use svm_common::Address;

    macro_rules! app_kv {
        ($app_addr:expr) => {{
            use std::{cell::RefCell, rc::Rc};

            use crate::app::AppKVStore;
            use crate::kv::FakeKV;

            use svm_kv::traits::KVStore;

            let raw_kv: Rc<RefCell<dyn KVStore>> = Rc::new(RefCell::new(FakeKV::new()));
            AppKVStore::new($app_addr, &raw_kv)
        }};
    }

    macro_rules! assert_vars {
        ($app:expr, $($var_id:expr => $expected:expr), *) => {{
            $(
                let actual = $app.read_var(VarId($var_id));
                assert_eq!(actual, $expected);
             )*
        }};
    }

    macro_rules! write_vars {
        ($app:expr, $($var_id:expr => $value:expr), *) => {{
            $(
                $app.write_var(VarId($var_id), $value.to_vec());
             )*
        }};
    }

    #[test]
    fn app_vars_are_persisted_on_commit() {
        // `var #0` consumes 4 bytes
        // `var #1` consumes 2 bytes
        let layout: DataLayout = vec![4, 2].into();
        let addr = Address::of("my-app");
        let kv = app_kv!(addr);

        // we create clones for later
        let layout_clone2 = layout.clone();
        let layout_clone3 = layout.clone();
        let kv_clone2 = Rc::clone(&kv.raw_kv);
        let kv_clone3 = Rc::clone(&kv.raw_kv);

        let mut app = AppStorage::new(layout, kv);

        // vars are initialized with zeros
        assert_vars!(app, 0 => [0, 0, 0, 0], 1 => [0, 0]);

        write_vars!(app, 0 => [10, 20, 30, 40], 1 => [50, 60]);

        // vars latest version are in memory
        assert_vars!(app, 0 => [10, 20, 30, 40], 1 => [50, 60]);

        // spin a new app with no in-memory dirty data
        let addr = Address::of("my-app");
        let kv2 = AppKVStore::new(addr, &kv_clone2);
        let app2 = AppStorage::new(layout_clone2, kv2);
        assert_vars!(app2, 0 => [0, 0, 0, 0], 1 => [0, 0]);

        // now, we'll persist `app` dirty changes
        app.commit();

        // we'll spin a new app with no caching
        let addr = Address::of("my-app");
        let kv3 = AppKVStore::new(addr, &kv_clone3);
        let mut app3 = AppStorage::new(layout_clone3, kv3);
        write_vars!(app3, 0 => [10, 20, 30, 40], 1 => [50, 60]);
    }

    #[test]
    #[cfg(debug_assertions)]
    #[should_panic]
    fn write_var_value_should_match_layout_length() {
        // `var #0` consumes 4 bytes
        let layout: DataLayout = vec![4].into();
        let addr = Address::of("my-app");
        let kv = app_kv!(addr);

        let mut app = AppStorage::new(layout, kv);
        app.write_var(VarId(0), vec![0, 0]);
    }
}
