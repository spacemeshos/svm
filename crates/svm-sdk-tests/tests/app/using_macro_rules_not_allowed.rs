use svm_sdk::app;

#[app]
mod App {
    macro_rules! print {
        () => {};
    }
}

fn main() {}
