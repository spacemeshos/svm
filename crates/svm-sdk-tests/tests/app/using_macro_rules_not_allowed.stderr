error: declaring `macro_rules!` inside `#[app]` is not supported.
 --> $DIR/using_macro_rules_not_allowed.rs:3:1
  |
3 | #[app]
  | ^^^^^^
  |
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
